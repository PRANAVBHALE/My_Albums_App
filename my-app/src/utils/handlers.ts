
import '@testing-library/jest-dom'
import { fetch, Headers, Request, Response } from 'cross-fetch'
import { rest } from 'msw'
import { setupServer } from 'msw/node'
import { UsersList } from './mocks/UsersList.mock'
import { AlbumsList } from './mocks/AlbumsList.mock'

global.fetch = fetch
global.Headers = Headers
global.Request = Request
global.Response = Response
global.matchMedia = global.matchMedia || function () {
    return {
      addListener: jest.fn(),
      removeListener: jest.fn(),
    };
  };

export const handlers = [
  rest.get('https://jsonplaceholder.typicode.com/albums?_start=0&_limit=5', (_req, res, ctx) => {
    return res(ctx.json(AlbumsList))
  }),

  rest.get('https://jsonplaceholder.typicode.com/users', (_req, res, ctx) => {
    return res(ctx.json(UsersList))
  }),

]

export const server = setupServer(...handlers)

// Enable the API mocking before tests.
beforeAll(() => server.listen())

// Reset any runtime request handlers we may add during the tests.
afterEach(() => server.resetHandlers())

// Disable the API mocking after the tests finished.
afterAll(() => server.close())