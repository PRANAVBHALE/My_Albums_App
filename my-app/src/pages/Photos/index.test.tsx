import React from "react";
import Photo from ".";
import { screen, waitFor } from "@testing-library/react";

import { server } from "../../utils/handlers";

import { renderWithProviders } from "../../utils/testUtils";
import { MemoryRouter } from "react-router-dom";
import { rest } from "msw";

jest.mock("react-router-dom", () => ({
  ...jest.requireActual("react-router-dom"),
  useParams: () => ({
    companyId: "company-id1",
    teamId: "team-id1",
  }),
}));

describe("Photo component", () => {
  it("should match snapshot", () => {
    let view = renderWithProviders(
      <MemoryRouter
        initialEntries={[
          {
            pathname: "/photo/album/1/0/5",
            search: "",
            state: {
              record: {
                albumId: 1,
                id: 1,
                imgUrl: "https://via.placeholder.com/150/e743b",
                name: "Pranav",
                title: "Album Title",
              },
            },
          },
        ]}
      >
        <Photo />
      </MemoryRouter>
    );

    expect(view).toMatchSnapshot();
  });

  it("should display Loader", () => {
    renderWithProviders(
      <MemoryRouter
        initialEntries={[
          {
            pathname: "/photo/album/1/0/5",
            search: "",
            state: {
              record: {
                albumId: 1,
                id: 1,
                imgUrl: "https://via.placeholder.com/150/e743b",
                name: "Pranav",
                title: "Album Title",
              },
            },
          },
        ]}
      >
        <Photo />
      </MemoryRouter>
    );

    const loader = screen.getByTestId("loader");
    expect(loader).toBeInTheDocument();
  });

  it("should display error", async () => {
    server.use(
      rest.get(
        `${process.env.REACT_APP_ENDPOINT}xxxxxxxxxxx/*`,
        (_req, res, ctx) => {
          return res(ctx.status(500), ctx.json("an error has occurred"));
        }
      )
    );

    const { getByText } = renderWithProviders(
      <MemoryRouter
        initialEntries={[
          {
            pathname: "/photo/album/1/0/5",
            search: "",
            state: {
              record: {
                albumId: 1,
                id: 1,
                imgUrl: "https://via.placeholder.com/150/e743b",
                name: "Pranav",
                title: "Album Title",
              },
            },
          },
        ]}
      >
        <Photo />
      </MemoryRouter>
    );

    const loader = screen.getByTestId("loader");
    expect(loader).toBeInTheDocument();

    await waitFor(() => {
      const error = getByText("Opps!!! Something went wrong!");
      expect(error).toBeInTheDocument();
    });
  });

  it("should success", async () => {
    const { getByText, getByAltText, queryByText, getByTestId } =
      renderWithProviders(
        <MemoryRouter
          initialEntries={[
            {
              pathname: "/photo/album/1/0/5",
              search: "",
              state: {
                record: {
                  albumId: 1,
                  id: 1,
                  imgUrl: "https://via.placeholder.com/150/e743b",
                  name: "Pranav",
                  title: "Album Title",
                },
              },
            },
          ]}
        >
          <Photo />
        </MemoryRouter>
      );

    const loader = screen.getByTestId("loader");
    expect(loader).toBeInTheDocument();

    await waitFor(() => {
      const table = getByTestId("photo-table");
      expect(table).toBeInTheDocument();
    });
  });

  it("should have owner name and title column", async () => {
    const { getByText, getByAltText, queryByText, getByTestId, getByRole } =
      renderWithProviders(
        <MemoryRouter
          initialEntries={[
            {
              pathname: "/photo/album/1/0/5",
              search: "",
              state: {
                record: {
                  albumId: 1,
                  id: 1,
                  imgUrl: "https://via.placeholder.com/150/e743b",
                  name: "Pranav",
                  title: "Album Title",
                },
              },
            },
          ]}
        >
          <Photo />
        </MemoryRouter>
      );

    const loader = screen.getByTestId("loader");
    expect(loader).toBeInTheDocument();

    await waitFor(() => {
      const ownerAndTitle = getByText(
        "Owner - Pranav and his album - Album Title"
      );
      expect(ownerAndTitle).toBeInTheDocument();
    });
  });
});
