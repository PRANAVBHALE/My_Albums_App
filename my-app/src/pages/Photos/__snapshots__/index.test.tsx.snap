// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Photo component should match snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        data-testid="loader"
        style="display: flex; align-items: center; justify-content: center; height: 100vh;"
      >
        <div
          class="ant-space css-dev-only-do-not-override-15rg2km ant-space-horizontal ant-space-align-center"
        >
          <div
            class="ant-space-item"
          >
            <div
              aria-busy="true"
              aria-live="polite"
              class="ant-spin ant-spin-lg ant-spin-spinning css-dev-only-do-not-override-15rg2km"
            >
              <span
                class="ant-spin-dot ant-spin-dot-spin"
              >
                <i
                  class="ant-spin-dot-item"
                />
                <i
                  class="ant-spin-dot-item"
                />
                <i
                  class="ant-spin-dot-item"
                />
                <i
                  class="ant-spin-dot-item"
                />
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      data-testid="loader"
      style="display: flex; align-items: center; justify-content: center; height: 100vh;"
    >
      <div
        class="ant-space css-dev-only-do-not-override-15rg2km ant-space-horizontal ant-space-align-center"
      >
        <div
          class="ant-space-item"
        >
          <div
            aria-busy="true"
            aria-live="polite"
            class="ant-spin ant-spin-lg ant-spin-spinning css-dev-only-do-not-override-15rg2km"
          >
            <span
              class="ant-spin-dot ant-spin-dot-spin"
            >
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
              <i
                class="ant-spin-dot-item"
              />
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "store": Object {
    "@@observable": [Function],
    "dispatch": [Function],
    "getState": [Function],
    "replaceReducer": [Function],
    "subscribe": [Function],
  },
  "unmount": [Function],
}
`;
